File: 96_numTrees.py
class Solution:
    def numTrees(self, n: int) -> int:
        if n == 0:
            return 0

        def getAns(start: int, end: int):
            num = 0
            if start >= end:  # 只有一个数字的时候
                return 1

            i = start
            while i <= end:
                tree_lefts_num = getAns(start, i - 1)
                tree_rights_num = getAns(i + 1, end)

                num += tree_lefts_num * tree_rights_num

                i += 1
            return num

        return getAns(1, n)


def main():
    solution = Solution()
    print(solution.numTrees(3))


if __name__ == '__main__':
    main()

File: leetcode_vs/leetcode_vs/96_numTrees.cpp
class Solution {
public:
    //动态规划
    int numTrees(int n) {
        vector<int> dp(n+1,0);
        dp[0] = 1;
        dp[1] = 1;
        //dp[n]等于分别以1到n为根节点的二叉树数目的和，当以i为根节点时，其种类为左子树的种类*右子树的种类
        //左子树节点个数  右节点个数
        //    0            i-1
        //    1            i-2
        //    ....         ....
        //    i-1           0
        for(int i=2;i<=n;i++)
        {
            for(int j = 0;j<i;j++)
                dp[i] += dp[j]*dp[i-j-1];
        }
        return dp[n];
    }
};